File: FateGrandAutomata/CardPriorityActivity.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Android.App;
using Android.OS;
using AndroidX.AppCompat.App;
using AndroidX.Preference;
using AndroidX.RecyclerView.Widget;

namespace FateGrandAutomata
{
    [Activity(Label = "Card Priority")]
    public class CardPriorityActivity : AppCompatActivity, IOnStartDragListener
    {
        ItemTouchHelper _itemTouchHelper;

        protected override void OnCreate(Bundle SavedInstanceState)
        {
            base.OnCreate(SavedInstanceState);
            SetContentView(Resource.Layout.card_priority);

            var preferences = PreferenceManager.GetDefaultSharedPreferences(this);
            var key = GetString(Resource.String.pref_card_priority);
            var cardPriority = preferences.GetString(key, DefaultCardPriority);

            // Handle simple mode and empty string
            if (cardPriority.Length == 3 || string.IsNullOrWhiteSpace(cardPriority))
            {
                cardPriority = DefaultCardPriority;
            }

            _cardScores = Card.GetCardScores(cardPriority).ToList();

            var adapter = new RecyclerListAdapter(_cardScores, this);

            var recyclerView = FindViewById<RecyclerView>(Resource.Id.card_priority_lv);
            recyclerView.HasFixedSize = true;
            recyclerView.SetAdapter(adapter);
            recyclerView.SetLayoutManager(new LinearLayoutManager(this));

            var callback = new ItemTouchHelperCallback(adapter);
            _itemTouchHelper = new ItemTouchHelper(callback);
            _itemTouchHelper.AttachToRecyclerView(recyclerView);
        }

        static string FilterCapitals<T>(T Obj)
        {
            var s = Obj.ToString();

            var sb = new StringBuilder();

            foreach (var t in s)
            {
                if (char.IsUpper(t))
                    sb.Append(t);
            }

            return sb.ToString();
        }

        protected override void OnPause()
        {
            base.OnPause();

            if (_cardScores == null)
                return;

            var val = _cardScores
                .Select(FilterCapitals)
                .Aggregate((Result, Current) => $"{Result}, {Current}");

            var preferences = PreferenceManager.GetDefaultSharedPreferences(this);
            var key = GetString(Resource.String.pref_card_priority);
            preferences.Edit()
                .PutString(key, val)
                .Commit();
        }

        List<CardScore> _cardScores;

        const string DefaultCardPriority = "WB, WA, WQ, B, A, Q, RB, RA, RQ";
        public void OnStartDrag(RecyclerView.ViewHolder ViewHolder)
        {
            _itemTouchHelper?.StartDrag(ViewHolder);
        }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/ui/card_priority/CardPriorityActivity.kt
package com.mathewsachin.fategrandautomata.ui.card_priority

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.edit
import androidx.recyclerview.widget.LinearLayoutManager
import com.mathewsachin.fategrandautomata.R
import com.mathewsachin.fategrandautomata.scripts.CardScore
import com.mathewsachin.fategrandautomata.scripts.modules.cardPriorityStageSeparator
import com.mathewsachin.fategrandautomata.scripts.modules.getCardScores
import com.mathewsachin.fategrandautomata.scripts.prefs.defaultCardPriority
import com.mathewsachin.fategrandautomata.scripts.prefs.getStringPref
import com.mathewsachin.fategrandautomata.ui.AutoSkillItemActivity
import kotlinx.android.synthetic.main.card_priority.*

fun String.filterCapitals(): String {
    return this
        .asSequence()
        .filter { it.isUpperCase() }
        .joinToString(separator = "")
}

class CardPriorityActivity : AppCompatActivity() {
    private lateinit var cardScores: MutableList<MutableList<CardScore>>

    private lateinit var autoSkillPref: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.card_priority)

        val autoSkillKey = intent.getStringExtra(AutoSkillItemActivity::autoSkillItemKey.name)
            ?: throw IllegalArgumentException("Missing AutoSkill item key in intent")
        autoSkillPref = getSharedPreferences(autoSkillKey, Context.MODE_PRIVATE)

        var cardPriority = getStringPref(R.string.pref_card_priority, Prefs = autoSkillPref)

        // Handle simple mode and empty string
        if (cardPriority.length == 3 || cardPriority.isBlank())
        {
            cardPriority = defaultCardPriority
        }

        cardScores = cardPriority
            .splitToSequence(cardPriorityStageSeparator)
            .map { getCardScores(it).toMutableList() }
            .toMutableList()

        val adapter = CardPriorityListAdapter(cardScores)

        val recyclerView = card_priority_list
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this)

        card_priority_add_btn.setOnClickListener {
            cardScores.add(mutableListOf<CardScore>().apply {
                addAll(cardScores[0])
            })

            adapter.notifyItemInserted(cardScores.lastIndex)
        }

        card_priority_rm_btn.setOnClickListener {
            if (cardScores.size > 1) {
                cardScores.removeAt(cardScores.lastIndex)

                adapter.notifyItemRemoved(cardScores.lastIndex + 1)
            }
        }
    }

    override fun onPause() {
        super.onPause()

        val value = cardScores.joinToString(cardPriorityStageSeparator) {
            it.joinToString { m -> m.toString().filterCapitals() }
        }

        val key = getString(R.string.pref_card_priority)
        autoSkillPref.edit(commit = true) { putString(key, value) }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.card_priority_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_card_priority_info -> {
                AlertDialog.Builder(this)
                    .setMessage("W: Weak (Effective)\nR: Resistive\n\nB: Buster\nA: Arts\nQ: Quick")
                    .setTitle("Info")
                    .setPositiveButton(android.R.string.yes, null)
                    .show()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
